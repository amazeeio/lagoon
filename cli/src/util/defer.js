// @flow

type Defer<R> = {
  promise: Promise<R>,
  resolve: (result: Promise<R> | R) => void,
  reject: (error: any) => void,
};

/**
 * Creates a defer object, which is useful to combine
 * Promises with nodelike callback interfaces
 */
export default function defer(): Defer<*> {
  const deferred = {};

  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve;
    deferred.reject = reject;
  });

  return deferred;
}
