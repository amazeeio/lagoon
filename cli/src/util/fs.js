// @flow

/**
 * This module exists to promisify all the fs functionality
 */

import fs from 'fs';
// Polyfill for Node 8's util.promisify
import promisify from 'util.promisify';

import type { Stats } from 'fs';

type ReadFileFn = (filename: string, enc?: string) => Promise<Buffer | string>;
type WriteFileFn = (
  filename: string,
  data: Buffer | string,
  options?: Object | string,
) => Promise<void>;
type UnlinkFn = (filename: string) => Promise<void>;

export const readFile: ReadFileFn = promisify(fs.readFile);
export const writeFile: WriteFileFn = promisify(fs.writeFile);
export const unlink: UnlinkFn = promisify(fs.unlink);

// TODO: @ryyppy Can we use something simpler like promisify(fs.access)?
export async function fileExists(path: string): Promise<boolean> {
  try {
    const stats: Stats = await promisify(fs.lstat)(path);

    if (stats.isFile()) {
      return true;
    }

    return false;
  } catch (err) {
    return false;
  }
}
