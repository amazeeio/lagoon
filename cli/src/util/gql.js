// @flow

// A no-op tagged template function to signal that queries are GraphQL for tooling
export default function gql(
  literals: Array<string>,
  ...substitutions: Array<any>
) {
  let result = '';

  // run the loop only for the substitution count
  substitutions.forEach((substitution, key) => {
    result += literals[key];
    result += substitution;
  });

  // add the last literal
  result += literals[literals.length - 1];

  return result;
}
