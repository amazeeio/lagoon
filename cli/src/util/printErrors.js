// @flow

import { red } from 'chalk';
import R from 'ramda';

type Cerr = typeof console.error;
type LaguError = { +message: string, +stack?: string };

export function printErrors(
  cerr: Cerr,
  ...errors: Array<LaguError | string>
): number {
  R.compose(
    R.forEach(cerr),
    R.map(red),
    R.map(
      err =>
        typeof err === 'string'
          ? err
          : R.propOr(
            // ...otherwise, use the message property as a fallback
            R.prop('message', err),
            // Try to get the stack...
            'stack',
            err,
          ),
    ),
  )(errors);
  return 1;
}

export function printNoConfigError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      '.lagoon.yml config file not found. Please create one with "lagoon init".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printProjectConfigurationError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      'No project configured. Please create a .lagoon.yml config file with "lagoon init" or pass a project to this command via the --project option.\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printGraphQLErrors(
  cerr: Cerr,
  ...errors: Array<LaguError>
): number {
  if (
    R.find(R.propEq('message', 'Unauthorized - Bearer Token Required'))(errors)
  ) {
    return printErrors(cerr, {
      message:
        'No authentication token found. Please log in first with "lagoon login".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-login',
    });
  }
  const errorsPluralized = R.length(errors) === 1 ? 'an error' : 'errors';
  return printErrors(
    cerr,
    `Oops! The Lagoon API returned ${errorsPluralized}:`,
    ...errors,
  );
}
