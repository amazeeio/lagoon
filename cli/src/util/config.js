// @flow

import { fileExists, writeFile } from './fs';
import yaml from 'js-yaml';

import path from 'path';
import co from 'co';

type DeployTask = {
  before_deploy: Array<string>,
  after_deploy: Array<string>,
};

// TODO: Type the rest of the config
export type AmazeeConfig = {
  sitegroup: string,
  deploy_tasks: {
    [name: string]: DeployTask,
  },
};

export type AmazeeConfigInput = {
  sitegroup: string,
};

export function createConfig(
  filepath: string,
  inputOptions: AmazeeConfigInput,
): Promise<void> {
  const config: AmazeeConfig = {
    ...inputOptions,
    deploy_tasks: {
      task1: {
        before_deploy: [],
        after_deploy: [],
      },
    },
  };
  const yamlConfig = yaml.safeDump(config);
  return writeFile(filepath, yamlConfig);
}

function* walker(
  dir: string,
  filename: string,
  root: string,
): Generator<*, ?string, *> {
  let next = dir;

  while (true) {
    const full = path.join(next, filename);
    const exists = yield fileExists(full);

    if (exists) {
      return full;
    } else if (next === root) {
      // ${root}/${filename} does not exist, quit
      return null;
    }

    // Otherwise go up one level
    next = path.dirname(next);
  }
}

/**
 * Tries to resolve a given filename, starting at cwd.
 * Will try ${cwd}/$filename, if not found it will step up
 * one directory (..) until it reaches root (/${filename}).
 *
 * If no file was found, it will return null, otherwise the path
 * of the found file.
 */
export async function findConfig(
  filename: string,
  cwd: string,
): Promise<?string> {
  const start = path.join(cwd, filename);
  const { root } = path.parse(start);

  return co(walker(cwd, filename, root));
}

export function parseConfig(yamlContent: string): AmazeeConfig {
  // TODO: Add schema validation in there if necessary
  return yaml.safeLoad(yamlContent);
}
