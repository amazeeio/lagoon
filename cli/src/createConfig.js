// @flow

import { writeFile } from './util/fs';
import yaml from 'js-yaml';
import type { AmazeeConfig, AmazeeConfigInput } from './parseConfig';

export default function createConfig(
  filepath: string,
  inputOptions: AmazeeConfigInput,
): Promise<void> {
  const config: AmazeeConfig = {
    ...inputOptions,
    deploy_tasks: {
      task1: {
        before_deploy: [],
        after_deploy: [],
      },
    },
  };
  const yamlConfig = yaml.safeDump(config);
  return writeFile(filepath, yamlConfig);
}
