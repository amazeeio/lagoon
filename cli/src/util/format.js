// @flow

import R from 'ramda';
import {
  FORMAT,
  FORMAT_CHOICE_TABLE,
  FORMAT_CHOICE_SIMPLE,
  FORMAT_CHOICE_JSON,
  FORMAT_CHOICE_CSV,
} from '../config/globalOptions';
import { getConfig } from '../config';

import type { FormatData } from '../types/Format';

export default function format(data: FormatData) {
  return R.cond([
    [
      R.equals(FORMAT_CHOICE_TABLE),
      () => require('../formatters/table').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_SIMPLE),
      () => require('../formatters/simple').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_JSON),
      () => require('../formatters/json').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_CSV),
      () => require('../formatters/csv').default(data),
    ],
  ])(R.prop(FORMAT, getConfig()));
}
