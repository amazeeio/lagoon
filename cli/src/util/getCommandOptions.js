// @flow

import R from 'ramda';
import { globalOptions } from '../config/globalOptions';

import type { Config } from '../types/Config';

export function getCommandOptions<CommandOptionsT: { [key: string]: any }>({
  config,
  commandOptions,
}: {|
  config: Config,
  commandOptions: CommandOptionsT,
|}): { [key: $Keys<CommandOptionsT>]: any, ...$Exact<Config> } {
  // Only return the properties that match the command options
  return R.pick(
    R.concat(R.keys(globalOptions), R.keys(commandOptions)),
    config,
  );
}
