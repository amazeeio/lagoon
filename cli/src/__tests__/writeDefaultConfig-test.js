// @flow

import createConfig from '../createConfig';
import { writeFile } from '../util/fs';

jest.mock('../util/fs');

function _mock(fn: any): JestMockFn {
  return fn;
}

describe('createConfig', () => {
  it('should write default config to given path', async () => {
    createConfig('amazeeio.yml', { sitegroup: 'your_sitegroup' });

    const [filename, data] = _mock(writeFile).mock.calls[0];

    expect(filename).toBe('amazeeio.yml');
    expect(data).toMatchSnapshot();
  });
});
