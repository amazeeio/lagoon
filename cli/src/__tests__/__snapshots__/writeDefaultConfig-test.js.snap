// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfig should write default config to given path 1`] = `
"sitegroup: your_sitegroup
deploy_tasks:
  task1:
    before_deploy: []
    after_deploy: []
"
`;
